Allow custom error code in throwIf

rschu1ze opened the pull request:
Fixes #39793

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
It is now possible to set a custom error code for the exception thrown by function "throwIf"

rschu1ze committed:
style: improve code aesthetics

rschu1ze committed:
feat: allow custom error code for SQL function throwIf()

yakov-olkhovskiy commented on docs/ru/sql-reference/functions/other-functions.md:
```diff
@@ -1727,10 +1727,13 @@ SELECT joinGet(db_test.id_val,'val',toUInt32(number)) from numbers(4) SETTINGS j
 Бросает исключение, если аргумент не равен нулю.
-custom_message - необязательный параметр, константная строка, задает текст сообщения об ошибке.
+`custom_message` - необязательный параметр, константная строка, задает текст сообщения об ошибке.
+`error_code` - необязательный параметр, константое число, задает устанавливает код ошибки.
```
```suggestion
`error_code` - необязательный параметр, константое число, задает код ошибки.
```

yakov-olkhovskiy committed:
Update 00602_throw_if.reference

yakov-olkhovskiy committed:
Update docs/ru/sql-reference/functions/other-functions.md

rschu1ze committed:
style: "Fix" style check in CI

yakov-olkhovskiy commented on src/Functions/throwIf.cpp:
```diff
@@ -22,6 +22,11 @@ namespace ErrorCodes
 namespace
 {
 
+/// The regex-based code style check script in CI complains when it sees "ErrorCodes:: ErrorCode" (space added to avoid another match).
+/// Because this expression is only used in this file, don't add some suppression mechanism to the already complex style checker, instead
+/// work around by creating a namespace alias.
+namespace ErrorCodeAlias = ErrorCodes;
+
```
I was wondering what da heck style check is complaining about :)

rschu1ze replied:
It says that "ErrorCodes::E" is used but not defined :grimacing: ... It's basically a shortcoming of being based on regexes.

tavplubix commented on tests/queries/0_stateless/00602_throw_if.sh:
```diff
@@ -4,9 +4,11 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
-default_exception_message="Value passed to 'throwIf' function is non zero"
+default_exception_message="Value passed to 'throwIf' function is non-zero"
 custom_exception_message="Number equals 1000000"
+custom_error_code="42"
```
Let's also add tests with some special error codes that can be processed special way (like `UNEXPECTED_PACKET_FROM_CLIENT` in `TCPHandler` or `OK`) and error codes that does not exist (like `123456` or `-1`)

rschu1ze replied:
Sure, will do.

rschu1ze committed:
fix: prevent possible race between getReturnTypeImpl() and executeImpl()

I don't know if the context (in detail: the config settings) can change
between calls to getReturnTypeImpl() and executeImpl() but we better
don't find out. Instead, read setting "allow_custom_error_code_in_throwif"
just just once in the ctor of FunctionThrowIf.

rschu1ze committed:
Only allow 8/16/32-bit signed integers as custom error code data type

The internally used data type for error code is "int", and by
disallowing unsigned integers or integers >= 64 bit we avoid truncation
issues.

rschu1ze committed:
test: extend tests for function throwIf() with edge cases

rschu1ze commented:
Test errors look unrelated.

tavplubix commented:
AST fuzzer (MSan) - looks like https://github.com/ClickHouse/ClickHouse/issues/40164
Integration tests (asan) [1/3] - test_concurrent_backups_s3/test.py::test_concurrent_backups - @vitlibar, please take a look
Stateless tests (aarch64) - 00993_system_parts_race_condition_drop_zookeeper - https://github.com/ClickHouse/ClickHouse/pull/40246
